### PR TITLE
citystate

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

**Review by bqi343 — APPROVED** [review]

File: solutions/usaco-667.mdx [hunk 1]
```diff
@@ -2,65 +2,150 @@
...
+<LanguageSection>
 <CPPSection>
 
-As the editorial mentions, we can store the number of times a given four-letter
-string appears in a `map`.
+<Info>
 
-**Time Complexity:** $\mathcal{O}(N\log N)$ with `map`, $\mathcal{O}(N)$ with
-`unordered_map`.
+As the below solution uses `map`, the complexity is actually
```
Comment (jessechoe10) [inline solutions/usaco-667.mdx:20]:
```suggestion
The time complexity of the below solution is actually
```

File: solutions/usaco-667.mdx [hunk 2]
```diff
@@ -69,21 +154,110 @@ int main() {
...
 		string state;
 		read >> city >> state;
-		city = city.substr(0,2);
-		if (city != state) {
-			total += seen[index(state + city)];
+		city = city.substr(0, 2);  // We only care about the first two letters of the city
+		pairs.push_back({city, state});
+	}
+
+	vector<int> seen(26 * 26 * 26 * 26);
```
Comment (jessechoe10) [inline solutions/usaco-667.mdx:167]:
```suggestion
	vector<int> seen(MAXN);
```
Make ```MAXN``` a global variable and set it equal to `26 * 26 * 26 * 26` or `pow(26, 4)`. Generally, it's a good idea to put constants as global variables...

File: solutions/usaco-667.mdx [hunk 3]
```diff
@@ -69,21 +154,110 @@ int main() {
...
+
+
+pairs = []
+with open('citystate.in') as read:
+	for _ in range(int(read.readline())):
+		city, state = read.readline().strip().split()
+		city = city[:2]  # We only care about the first two letters of the city
+		pairs.append((city, state))
+
+seen = [0 for _ in range(26 ** 4)]
```
Comment (jessechoe10) [inline solutions/usaco-667.mdx:252]: i guess here too??

**Review by jessechoe10 — CHANGES_REQUESTED** [review]
some of them are optional but i think they'd look/sound much better if you did these

File: solutions/usaco-667.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/usaco-667.mdx:20]: i feel like mentioning `map` is necessary

Comment (jessechoe10) [reply solutions/usaco-667.mdx:20]: uhh i said "since it uses `map`" the following line

**Review by ryanchou-dev — APPROVED** [review]

File: solutions/usaco-667.mdx [hunk 4]
```diff
@@ -2,65 +2,150 @@
...
+		for (int p = 0; p < pairNum; p++) {
+			StringTokenizer cityState = new StringTokenizer(read.readLine());
+			String city = cityState.nextToken();
+			String state = cityState.nextToken();
+			// We only care about the first two letters of the city
+			city = city.substring(0, 2);
+			pairs[p] = new Pair(city, state);
+		}
+
+		HashMap<String, Integer> seen = new HashMap<>();
```
Comment (nathangong) [inline solutions/usaco-667.mdx:87]:
```suggestion
		Map<String, Integer> seen = new HashMap<>();
```

**Review by nathangong — APPROVED** [review]

**Pull request merged**